### DISCUSSION TITLE
Get content of random file for given language?

### DISCUSSION THREAD

**Discussion by drewgingerich** [community/community#24597] [Apps, API and Webhooks]
<p>I want to fetch the contents of a random file written in a given programming language to use for a typing game. Using JavaScript as an example, I’ve used the search API to find repositories containing JavaScript:</p>
<pre><code class="lang-auto">https://api.github.com/search/repositories?q=language:javascript&amp;sort=stars&amp;order=desc
</code></pre>
<p>But I’m having trouble figuring out how to list a repo’s JavaScript files and fetch their contents. Is there a way to do this? Is there a better way to get the contents of a random JavaScript file (preferrably from a highly-starred repo)?</p>
<p>Thanks for any help anyone can give!</p>

**Comment by jfullstackdev** [comment]
<p>how about this one ?</p>
<p><a href="https://docs.github.com/en/rest/repos/contents#get-repository-content">https://docs.github.com/en/rest/repos/contents#get-repository-content</a></p>
<p>do the search first then once you have a target file, read the content</p>

**Comment by drewgingerich** [comment]
<p>Hi <a class="mention" href="/jdevstatic">@jdevstatic</a> that’s helpful, thank you!</p>
<p>I’m still not sure how to find a file of a particular language in a repo. My search query above lists repos that have a file using the language somewhere, but I can’t figure out a way to list those files and their paths.</p>

**Comment by jfullstackdev** [comment]
<p>I think there is a similar question on how to find the list of supported language by GitHub and this is the particular repo, I think so,</p>
<p>you can also ask them by raising an issue</p>
<aside class="onebox allowlistedgeneric" data-onebox-src="https://github.com/github/linguist">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">

      <a href="https://github.com/github/linguist" target="_blank" rel="noopener">GitHub</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/345;"><img src="https://user-images.githubusercontent.com/47092464/181085555-4499f57d-890d-4a3e-8620-23c93fe28114.png" class="thumbnail" width="690" height="345" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/7/7/77893c16ec05e3ff4b48040974744ce73b97d732_2_690x345.png, https://aws1.discourse-cdn.com/github/optimized/3X/7/7/77893c16ec05e3ff4b48040974744ce73b97d732_2_1035x517.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/7/7/77893c16ec05e3ff4b48040974744ce73b97d732.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/7/7/77893c16ec05e3ff4b48040974744ce73b97d732_2_10x10.png"></div>

<h3><a href="https://github.com/github/linguist" target="_blank" rel="noopener">GitHub - github/linguist: Language Savant. If your repository's language is...</a></h3>

  <p>Language Savant. If your repository's language is being reported incorrectly, send us a pull request! - GitHub - github/linguist: Language Savant. If your repository's language is being rep...</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

**Comment by drewgingerich** [comment]
<p>I think there may be a miscommunication, I am not looking for a list of supported languages. What I want to do is:</p>
<ol>
<li>Get a list of repos containing an language (e.g. JavaScript)</li>
<li>For one of the listed repos, get a list of files written in that language</li>
<li>Get the content of one of the listed files</li>
</ol>
<p>I already know how to do step 1 using the search query in my original post. Your first comment showed me how to do step 3 (thanks again!). What I still don’t know is how to do step 2.</p>

**Comment by jfullstackdev** [comment]
<p>how about this one ?</p>
<aside class="onebox stackexchange" data-onebox-src="https://stackoverflow.com/questions/25022016/get-all-file-names-from-a-github-repo-through-the-github-api">
  <header class="source">

      <a href="https://stackoverflow.com/questions/25022016/get-all-file-names-from-a-github-repo-through-the-github-api" target="_blank" rel="noopener">stackoverflow.com</a>
  </header>

  <article class="onebox-body">
      <a href="https://stackoverflow.com/users/2980299/anton-antonov" target="_blank" rel="noopener">
    <img alt="Anton Antonov" src="https://user-images.githubusercontent.com/47092464/181085557-42017439-a2ce-4055-bd76-d00dce8fed56.png" class="thumbnail onebox-avatar" width="256" height="256">
  </a>

<h4>
  <a href="https://stackoverflow.com/questions/25022016/get-all-file-names-from-a-github-repo-through-the-github-api" target="_blank" rel="noopener">Get all file names from a Github repo through the Github API</a>
</h4>

<div class="tags">
  <strong>github, github-api</strong>
</div>

<div class="date">
  asked by
  
  <a href="https://stackoverflow.com/users/2980299/anton-antonov" target="_blank" rel="noopener">
    Anton Antonov
  </a>
  on <a href="https://stackoverflow.com/questions/25022016/get-all-file-names-from-a-github-repo-through-the-github-api" target="_blank" rel="noopener">06:02PM - 29 Jul 14 UTC</a>
</div>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>it includes the file name extension, so you would know the language</p>

**Comment by drewgingerich** [comment, marked as answer]
<p>Hey, thanks again for your help! I ended up using the search API to find the files as well. So my sequence is:</p>
<ol>
<li>Get a repository list from <code>api.github.com/search/repositories?q=language:${codeLanguage}&amp;sort=stars&amp;order=desc</code>
</li>
<li>Get a file list from <code>api.github.com/search/code?q=%20+language:${codeLanguage}+repo:${repoName}</code>
</li>
<li>Get file contents using URLs provided by step 2 (which use the repo contents API you showed me)</li>
</ol>
<p>This way I don’t have to crawl the Git tree to find files and can keep API requests few. The search also handles finding files by language even when the files may have one of multiple extensions (JavaScript has many).</p>
<p>One weird thing is that the Search API requires a string to search for, so I’ve just given it a space – that’s the <code>q=%20</code> – figuring almost all code files will have one somewhere!</p>
